File: py/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidHelper(self, node: TreeNode, lower, upper: int) -> bool:
        if not node: # base case, empty tree is a valid BST
            return True

        val = node.val
        if (val > lower and val < upper) and \
            self.isValidHelper(node.left, lower, node.val) and \
            self.isValidHelper(node.right, node.val, upper):
                return True
        return False


    def isValidBST(self, root: TreeNode) -> bool:
        return self.isValidHelper(root, float('-inf'), float('inf'))



class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        prev = None
        curr = root
        stack = []
        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.left
            s = stack.pop()
            if prev and s <= prev.val:
                return False
            prev = s
            curr = s.right
        return True

File: cpp/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MIN, LONG_MAX);
    }

private:
    bool helper(TreeNode* node, long lower, long upper) {
        if (!node) {
            return true;
        }

        long val = node->val;
        if (val <= lower or val >= upper) {
            return false;
        }

        if (!helper(node->right, val, upper) or !helper(node->left, lower, val)) {
            return false;
        }
        return true;
    }
};